### ISSUE TITLE
Namelist parsed successfully despite error when using in an application

### ISSUE THREAD

**Issue by ColemanTom** [marshallward/f90nml#137]
I have a namelist which I believe should fail to be read in.

```
 &Station Id='171',ObsType='23501'',
  RejRep='F',RejUV='F',
  ConfThres=70., ThinRound='SATW' /
```

When using this tool, it passes and is read in. I think it should fail because of the second single quote in the `ObsType` entry. If the second single quote is removed and added to the ThinRound entry, then it will correctly fail. But if its anwywhere before that last single quote, it passes.

**Comment by marshallward** [comment]
It is unusual, but it may still be a valid namelist.  What you have written is interpreted by `f90nml` as the following:
```
&station
    id = '171'
    obstype = "23501',  RejRep=", "F'"
    rejuv = 'F'
    confthres = 70.0
    thinround = 'SATW'
/
```
Repeated apostrophes escape to `'` in Fortran, becoming part of the string, which sets off a cascade of values which are nonsense to us, yet may still have a valid interpretation.

In this case, `f90nml` interprets the value of `obstype` as a 2-element array of strings with the following values:

* `23501', RejRep=`  (terminated after the first unescaped `'`)
* `F'` (a non-delimited string, i.e. without quotes, followed by an unpaired `'`, treated as part of the string)

After hitting the comma, followed by the next key (`rejuv`), it assumes that we've reached the end of the string and restarts the parsing process.

So I think one could say that it returned the correct result, or at least something very close, [from a certain point of view](https://duckduckgo.com/?q=from+a+certain+point+of+view+robot+chicken&t=ffab&iar=videos&iax=videos&ia=videos&iai=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DpSOBeD1GC_Y).

There are some potential errors here worth considering, although it is a bit beyond the scope of your question:

1. Should the endline (`\n`) be part of the value of `RejRep`?  I would have thought yes, but the rules of list-directed I/O may say otherwise here.  I would need to check.
2. No space was required between the string and `F`.  I think this is correct, but I am not sure and would need to consult the standard.
3. The inclusion of `'` in `F'` was also a surprise to me. I guess, as a non-delimited string, `f90nml` just appended it to the current string.  But I honestly have no idea what ought to happen here.

This namelist would certainly have failed runtime if `obstype` was not declared as a 2-element array.   Unfortunately, we don't have access to the original source code and sometimes have to make our best effort to interpret the file.

---

The second case you mentioned, where you moved this extra single quote to the final string, is presumably something like this.
```
 &Station Id='171',ObsType='23501'',
  RejRep='F',RejUV='F',
  ConfThres=70., ThinRound='SATW'' /
```
Here, `ThinRound` starts off as a string with value `SATW' /` but hits the end of the file before it can terminate the string.  I would say that this is the correct behavior.
----

To sum up, this is a very weird example, but it may still have a sensible interpretation.  So I would not yet assume that it should have failed.  But there are some interesting questions here worth investigating.

I will test some of these out in my available compilers, but I am not sure I would even trust the compilers to give the correct, standards-compliant result.  This would also need careful reading of the rules for list-directed IO.

**Comment by ColemanTom** [comment]
Thanks for the very quick reply. In this case, yes, ObsType is defined as a character type (e.g. `CHARACTER(len=..) :: ObsType`), not an array though, so that explains why it was failing in the application, but valid here. I learnt a lot from just reading your response.

Is there any way to tell `f90nml` what types things are to add some extra enforcement? I'm hoping to use `f90nml` to provide some quick validation through CI, and it has captured some errors I've seen in namelists I've been provided with, so will immediately add value as is, but having a way to strictly enforce types to capture this sort of issue would be great too.

**Comment by marshallward** [comment]
No, I don't think anything like that is currently possible.  I've thought that something like you suggest may be needed someday, but as yet never tried to implement anything close to it.

For a namelist like this
```
&grp
   s = 123
/
```
the `123` could be an integer, but it could also be a float or even a non-delimited string.  This problem is pretty close to your own, I think.

I could see some kind of strict mode where the keys are pre-assigned and the values are filled by the parser, with errors being raised if they do not agree.  But I also can't see myself finding the time to work on that anytime soon.

So no good news in the short term, but perhaps in the future.

**Comment by marshallward** [comment]
One possible solution is to provide an option which disables non-delimited strings and simply raising an error if one is encountered.

This is not a great solution, partly because Fortran processors are actually obliged to print namelists with non-delimited strings if unspecified (an enormous pet peeve of mine, btw](https://github.com/j3-fortran/fortran_proposals/issues/67), so such namelists are actually quite common.  So it would need to be disabled on default.

But non-delimited strings are generally a bad idea for many reasons, and disabling them would go a long way towards addressing your issue, I think.

**Comment by ColemanTom** [comment]
With my knowledge of these namelists, I think I can work past this on my end using a crude replacement, e.g.

```python
import f90nml
import re

with open("namelist") as f:
    namelist = f.read()
modified_namelist = re.sub(r"([a-zA-Z])=", r"\1(1)=", namelist)
f90nml.reads(modified_namelist)
```

It's very hacky, but I think it is safe for the namelists I'm looking at. Assigning values to array's in them are always done explcitly, never implied. For example the syntax is always along the lines of `value(1:2)='a', 'b'` for arrays.

**Comment by marshallward** [comment]
I think for now let's close this issue then.  A better solution is possible, but would be of larger scope and probably not any time soon.
